File: splitmerge/heuristics.py
"""Admissible heuristics for the split-merge machine.

All heuristics have the form ``h = h0 + 2 * (forced bounces)``, where a *bounce*
is a card forced to enter the deck more than once.  Each move handles one card,
so the number of moves is the base cost plus two per forced bounce
(docs/NOTES.md, §I.1 and §I.4).  Three heuristics, increasing strength:

    h0      base charge only: 2*(above-base deck) + |A| + |B|
    h_best  + 2*max((LIS-2)+ buried,  clique-2)      -- the "max" version
    h_joint + 2*(buried + OCT_pre)                   -- the joint version

``h_joint`` dominates ``h_best`` dominates ``h0`` everywhere, and all three are
admissible (verified by full BFS for n <= 7 in the test suite).
"""
import bisect
import itertools
from collections import deque
from functools import lru_cache

from .machine import base_len
from .oct import exact_oct


def _lis_len(seq):
    """Length of the longest strictly-increasing subsequence."""
    tails = []
    for x in seq:
        i = bisect.bisect_left(tails, x)
        if i == len(tails):
            tails.append(x)
        else:
            tails[i] = x
    return len(tails)


def _buried_set(A, B):
    """Cards that are *buried*: a smaller card sits below them in the same
    buffer, so they must lift off before it settles -> they must bounce
    (docs/NOTES.md, §I.4).
    """
    out = set()
    for buf in (A, B):
        for i, x in enumerate(buf):
            if any(y < x for y in buf[:i]):
                out.add(x)
    return out


def h0(state):
    """Base charge: each above-base deck card costs >= 2 (evacuate + return),
    each buffer card >= 1 (return).  Provably admissible; exact on rotations.
    """
    D, A, B = state
    k = base_len(D)
    return 2 * (len(D) - k) + len(A) + len(B)


def _clique_chain(D, A, B, k):
    """Size of the largest mutual-conflict clique, computed as the longest
    chain of the conflict comparability graph (edges orient smaller-value ->
    larger).  A clique of size c forces c-2 bounces (only 2 buffers).
    """
    # node = (value, kind, position); kind 0 = above-base deck (in pi order),
    # 1 = buffer A, 2 = buffer B.
    nodes = []
    pi = list(reversed(D[k:]))
    for pos, v in enumerate(pi):
        nodes.append((v, 0, pos))
    for d, v in enumerate(A):
        nodes.append((v, 1, d))
    for d, v in enumerate(B):
        nodes.append((v, 2, d))
    N = len(nodes)
    if N == 0:
        return 0

    def edge(i, j):
        vi, ki, xi = nodes[i]
        vj, kj, xj = nodes[j]
        if ki == 0 and kj == 0:                    # deck-deck: increasing in pi
            return (vi < vj) if xi < xj else (vj < vi)
        if ki == 0 and kj >= 1:                    # deck vs buffer: buffer smaller
            return vj < vi
        if kj == 0 and ki >= 1:
            return vi < vj
        if ki >= 1 and kj >= 1:                    # same buffer: deeper-smaller
            if ki != kj:
                return False
            return (vi < vj) if xi < xj else (vj < vi)
        return False

    order = sorted(range(N), key=lambda i: nodes[i][0])
    best = [1] * N
    ans = 0
    for a in range(N):
        i = order[a]
        for b in range(a):
            j = order[b]
            if edge(i, j) and best[j] + 1 > best[i]:
                best[i] = best[j] + 1
        ans = max(ans, best[i])
    return ans


@lru_cache(maxsize=None)
def h_best(state):
    """The "max" heuristic: ``h0 + 2*max((LIS-2)+buried, clique-2)``
    (docs/NOTES.md, §I.4)."""
    D, A, B = state
    k = base_len(D)
    pi = tuple(reversed(D[k:]))
    disjoint = max(0, _lis_len(pi) - 2) + len(_buried_set(A, B))
    clique = _clique_chain(D, A, B, k)
    return h0(state) + 2 * max(disjoint, max(0, clique - 2))


def _two_colourable(verts, adj, pre):
    """Is the graph ``(verts, adj)`` properly 2-colourable while respecting the
    fixed colours in ``pre``?  BFS propagation, seeding from the pre-coloured
    vertices first (so a free vertex never grabs a colour its pre-coloured
    neighbour will contradict)."""
    colour = {v: pre[v] for v in pre if v in set(verts)}
    dq = deque(colour.keys())
    while dq:                                  # propagate forced colours first
        u = dq.popleft()
        for w in adj[u]:
            if w in colour:
                if colour[w] == colour[u]:
                    return False
            else:
                colour[w] = 1 - colour[u]
                dq.append(w)
    for v in verts:                            # then free components
        if v in colour:
            continue
        colour[v] = 0
        dq = deque([v])
        while dq:
            u = dq.popleft()
            for w in adj[u]:
                if w in colour:
                    if colour[w] == colour[u]:
                        return False
                else:
                    colour[w] = 1 - colour[u]
                    dq.append(w)
    return True


def _soft_conflict_graph(state):
    """Build the soft-conflict graph (docs/NOTES.md, §I.4).

    Vertices: non-buried buffer cards (pre-coloured by their buffer) and all
    above-base deck cards (free).  A *soft* edge joins two cards that could go
    to different buffers but cannot share one if both are single-arrival:
    deck-deck increasing in pi, and deck ``d`` vs non-buried buffer card
    ``q < d``.  (Same-buffer pairs are handled separately by the buried count.)

    Returns ``(values, adjacency, pre_colour, n_buried)``.
    """
    D, A, B = state
    k = base_len(D)
    buried = _buried_set(A, B)
    pi = list(reversed(D[k:]))
    nodes = []                                 # (value, kind); kind 0=A, 1=B, 2=deck
    pre = {}
    for v in A:
        if v not in buried:
            nodes.append((v, 0))
            pre[v] = 0
    for v in B:
        if v not in buried:
            nodes.append((v, 1))
            pre[v] = 1
    for pos, v in enumerate(pi):
        nodes.append((v, 2))
    vals = [v for v, _ in nodes]
    adj = {v: set() for v in vals}
    # deck cards are appended in pi order; recover their pi position by index
    deck_pos = {}
    pos = 0
    for v, kind in nodes:
        if kind == 2:
            deck_pos[v] = pos
            pos += 1
    for i in range(len(nodes)):
        for j in range(i + 1, len(nodes)):
            vi, ki = nodes[i]
            vj, kj = nodes[j]
            if ki == 2 and kj == 2:
                e = (vi < vj) if deck_pos[vi] < deck_pos[vj] else (vj < vi)
            elif ki == 2 and kj < 2:
                e = vj < vi
            elif kj == 2 and ki < 2:
                e = vi < vj
            else:
                e = False
            if e:
                adj[vi].add(vj)
                adj[vj].add(vi)
    return vals, adj, pre, len(buried)


def _oct_pre(vals, adj, pre):
    """Minimum vertices to delete so the graph 2-colours respecting ``pre`` --
    the exact constrained OCT (see ``oct.exact_oct``), via odd-cycle
    branch-and-bound with the comparability chain bound (clique - 2 <= OCT)
    pruning the clique case.  Exact within a search budget; on a large
    far-from-clique graph it degrades gracefully to the admissible chain lower
    bound.  Validated against ``_oct_pre_bruteforce`` on all n <= 7 states.
    """
    value, _exact = exact_oct(vals, adj, pre)
    return value


def _oct_pre_bruteforce(vals, adj, pre):
    """Reference OCT by exhaustive deletion search -- the test oracle.  Correct
    but exponential; not used in production (see ``_oct_pre``)."""
    for d in range(len(vals) + 1):
        for drop in itertools.combinations(vals, d):
            dropped = set(drop)
            kept = [v for v in vals if v not in dropped]
            sub = {v: {w for w in adj[v] if w not in dropped} for v in kept}
            pre_kept = {v: pre[v] for v in kept if v in pre}
            if _two_colourable(kept, sub, pre_kept):
                return d
    return len(vals)


@lru_cache(maxsize=None)
def h_joint(state):
    """The joint heuristic: ``h0 + 2*(buried + OCT_pre)``
    (docs/NOTES.md, §I.4).  Eliminates the ``max`` of ``h_best`` by
    bounding the single-arrival set directly, and dominates ``h_best``.
    """
    vals, adj, pre, n_buried = _soft_conflict_graph(state)
    return h0(state) + 2 * (n_buried + _oct_pre(vals, adj, pre))

File: src/heuristics.rs
//! Admissible heuristics: `h0 <= h_best <= h_joint`, all lower bounds on the
//! optimal move count (`h = h0 + 2*(lower bound on forced bounces)`). See
//! docs/NOTES.md §I.4. `h_joint` reduces the bounce bound to a constrained OCT
//! on the soft-conflict graph (`oct::exact_oct`).

use std::collections::HashSet;

use crate::machine::{base_len, Card, State};
use crate::oct::{exact_oct, Graph, PreColour, Vertex, DEFAULT_BUDGET};

/// Length of the longest strictly-increasing subsequence.
pub fn lis_len(seq: &[Card]) -> usize {
    let mut tails: Vec<Card> = Vec::new();
    for &x in seq {
        let i = tails.partition_point(|&t| t < x); // bisect_left (cards distinct)
        if i == tails.len() {
            tails.push(x);
        } else {
            tails[i] = x;
        }
    }
    tails.len()
}

/// Cards that are *buried*: a smaller card sits below them in the same buffer.
fn buried_set(a: &[Card], b: &[Card]) -> HashSet<Card> {
    let mut out = HashSet::new();
    for buf in [a, b] {
        for i in 0..buf.len() {
            let x = buf[i];
            if buf[..i].iter().any(|&y| y < x) {
                out.insert(x);
            }
        }
    }
    out
}

/// Base charge: each above-base deck card costs >= 2, each buffer card >= 1.
pub fn h0(s: &State) -> u32 {
    let k = base_len(&s.d);
    (2 * (s.d.len() - k) + s.a.len() + s.b.len()) as u32
}

/// Above-base deck values read top-to-bottom (the forced first-departure order).
fn pi_seq(s: &State) -> Vec<Card> {
    let k = base_len(&s.d);
    s.d[k..].iter().rev().copied().collect()
}

/// Longest mutual-conflict clique, as the longest chain of the conflict
/// comparability graph (kind 0 = above-base deck in pi order, 1 = A, 2 = B).
fn clique_chain(s: &State) -> usize {
    let pi = pi_seq(s);
    let mut nodes: Vec<(Card, u8, usize)> = Vec::new();
    for (pos, &v) in pi.iter().enumerate() {
        nodes.push((v, 0, pos));
    }
    for (d, &v) in s.a.iter().enumerate() {
        nodes.push((v, 1, d));
    }
    for (d, &v) in s.b.iter().enumerate() {
        nodes.push((v, 2, d));
    }
    let n = nodes.len();
    if n == 0 {
        return 0;
    }
    let edge = |i: usize, j: usize| -> bool {
        let (vi, ki, xi) = nodes[i];
        let (vj, kj, xj) = nodes[j];
        if ki == 0 && kj == 0 {
            if xi < xj {
                vi < vj
            } else {
                vj < vi
            }
        } else if ki == 0 && kj >= 1 {
            vj < vi
        } else if kj == 0 && ki >= 1 {
            vi < vj
        } else if ki >= 1 && kj >= 1 {
            if ki != kj {
                false
            } else if xi < xj {
                vi < vj
            } else {
                vj < vi
            }
        } else {
            false
        }
    };
    let mut order: Vec<usize> = (0..n).collect();
    order.sort_by_key(|&i| nodes[i].0);
    let mut best = vec![1usize; n];
    let mut ans = 0;
    for a in 0..n {
        let i = order[a];
        for &j in &order[..a] {
            if edge(i, j) && best[j] + 1 > best[i] {
                best[i] = best[j] + 1;
            }
        }
        ans = ans.max(best[i]);
    }
    ans
}

/// The "max" heuristic: `h0 + 2*max((LIS-2)+buried, clique-2)`.
pub fn h_best(s: &State) -> u32 {
    let pi = pi_seq(s);
    let buried = buried_set(&s.a, &s.b);
    let disjoint = lis_len(&pi).saturating_sub(2) + buried.len();
    let clique = clique_chain(s);
    h0(s) + 2 * disjoint.max(clique.saturating_sub(2)) as u32
}

/// Build the soft-conflict graph: non-buried buffer cards (pre-coloured by their
/// buffer) and all above-base deck cards (free). Soft edges: deck-deck increasing
/// in pi, and deck `d` vs non-buried buffer card `q < d`. Returns
/// `(values, adjacency, pre_colour, n_buried)`.
pub fn soft_conflict_graph(s: &State) -> (Vec<Vertex>, Graph, PreColour, usize) {
    let buried = buried_set(&s.a, &s.b);
    let pi = pi_seq(s);
    // nodes: (value, kind) with kind 0 = A, 1 = B, 2 = deck
    let mut nodes: Vec<(Card, u8)> = Vec::new();
    let mut pre: PreColour = PreColour::new();
    for &v in &s.a {
        if !buried.contains(&v) {
            nodes.push((v, 0));
            pre.insert(v as Vertex, 0);
        }
    }
    for &v in &s.b {
        if !buried.contains(&v) {
            nodes.push((v, 1));
            pre.insert(v as Vertex, 1);
        }
    }
    for &v in &pi {
        nodes.push((v, 2));
    }
    let vals: Vec<Vertex> = nodes.iter().map(|&(v, _)| v as Vertex).collect();
    let mut adj: Graph = Graph::new();
    for &v in &vals {
        adj.entry(v).or_default();
    }
    // recover deck pi-position by appearance order among kind-2 nodes
    let mut deck_pos: std::collections::HashMap<Card, usize> = std::collections::HashMap::new();
    let mut pos = 0;
    for &(v, kind) in &nodes {
        if kind == 2 {
            deck_pos.insert(v, pos);
            pos += 1;
        }
    }
    for i in 0..nodes.len() {
        for j in (i + 1)..nodes.len() {
            let (vi, ki) = nodes[i];
            let (vj, kj) = nodes[j];
            let e = if ki == 2 && kj == 2 {
                if deck_pos[&vi] < deck_pos[&vj] {
                    vi < vj
                } else {
                    vj < vi
                }
            } else if ki == 2 && kj < 2 {
                vj < vi
            } else if kj == 2 && ki < 2 {
                vi < vj
            } else {
                false
            };
            if e {
                adj.get_mut(&(vi as Vertex)).unwrap().insert(vj as Vertex);
                adj.get_mut(&(vj as Vertex)).unwrap().insert(vi as Vertex);
            }
        }
    }
    (vals, adj, pre, buried.len())
}

/// The joint heuristic: `h0 + 2*(buried + OCT_pre)`; dominates `h_best`.
pub fn h_joint(s: &State) -> u32 {
    let (vals, adj, pre, n_buried) = soft_conflict_graph(s);
    let (oct, _) = exact_oct(&vals, &adj, &pre, DEFAULT_BUDGET);
    h0(s) + 2 * (n_buried + oct) as u32
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn known_compound_example() {
        // compound state from the docs: opt = 17
        let s = State {
            d: vec![6, 2, 5],
            a: vec![4, 1],
            b: vec![3],
        };
        assert_eq!(h0(&s), 9);
        assert_eq!(h_best(&s), 11);
        assert_eq!(h_joint(&s), 13); // joint is strictly tighter here
    }

    #[test]
    fn reversal_is_exact_for_strong_heuristics() {
        for n in 3..=13 {
            let s = State::reversed_deck(n);
            assert_eq!(h_best(&s), 4 * (n as u32 - 1));
            assert_eq!(h_joint(&s), 4 * (n as u32 - 1));
            assert_eq!(h0(&s), 2 * n as u32);
        }
    }
}
